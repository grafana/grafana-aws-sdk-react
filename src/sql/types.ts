import { DataQuery } from '@grafana/schema';
import { FillValueOptions } from './QueryEditor/FillValueSelect';

export const defaultKey = '__default';

export interface SQLQuery extends DataQuery {
  rawSQL: string;
  format?: number;
  fillMode?: { mode: FillValueOptions; value?: number };
}
