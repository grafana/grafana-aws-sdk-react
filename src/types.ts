import { DataSourceJsonData, DataSourceSettings } from '@grafana/data';

export enum AwsAuthType {
  Keys = 'keys',
  Credentials = 'credentials',
  Default = 'default', // was 'arn',
  EC2IAMRole = 'ec2_iam_role',
  /**
   * @deprecated use default
   */
  ARN = 'arn',
}

export interface AwsAuthDataSourceJsonData extends DataSourceJsonData {
  authType?: AwsAuthType;
  assumeRoleArn?: string;
  externalId?: string;
  profile?: string; // Credentials profile name, as specified in ~/.aws/credentials
  defaultRegion?: string; // region if it is not defined by your credentials file
  endpoint?: string;
}

export interface AwsAuthDataSourceSecureJsonData {
  accessKey?: string;
  secretKey?: string;
  sessionToken?: string;
}

export type AwsAuthDataSourceSettings = DataSourceSettings<AwsAuthDataSourceJsonData, AwsAuthDataSourceSecureJsonData>;
