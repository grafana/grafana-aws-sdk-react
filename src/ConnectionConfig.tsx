import React, { FC, useEffect, useMemo, useState } from 'react';
import { Input, Select, InlineField, ButtonGroup, ToolbarButton, FieldSet } from '@grafana/ui';
import {
  DataSourcePluginOptionsEditorProps,
  onUpdateDatasourceJsonDataOptionSelect,
  onUpdateDatasourceResetOption,
  onUpdateDatasourceJsonDataOption,
  onUpdateDatasourceSecureJsonDataOption,
} from '@grafana/data';
import { config } from '@grafana/runtime';
import { standardRegions } from './regions';
import { AwsAuthDataSourceJsonData, AwsAuthDataSourceSecureJsonData, AwsAuthType } from './types';
import { awsAuthProviderOptions } from './providers';

export const DEFAULT_LABEL_WIDTH = 28;
const toOption = (value: string) => ({ value, label: value });
const isAwsAuthType = (value: any): value is AwsAuthType => {
  return typeof value === 'string' && awsAuthProviderOptions.some((opt) => opt.value === value);
};
export interface ConnectionConfigProps<
  J extends AwsAuthDataSourceJsonData = AwsAuthDataSourceJsonData,
  S = AwsAuthDataSourceSecureJsonData
> extends DataSourcePluginOptionsEditorProps<J, S> {
  standardRegions?: string[];
  loadRegions?: () => Promise<string[]>;
  defaultEndpoint?: string;
  skipHeader?: boolean;
  skipEndpoint?: boolean;
  children?: React.ReactNode;
  labelWidth?: number;
}

export const ConnectionConfig: FC<ConnectionConfigProps> = (props: ConnectionConfigProps) => {
  const [regions, setRegions] = useState((props.standardRegions || standardRegions).map(toOption));
  const { loadRegions, onOptionsChange, skipHeader = false, skipEndpoint = false } = props;
  const { labelWidth = DEFAULT_LABEL_WIDTH, options } = props;
  let profile = options.jsonData.profile;
  if (profile === undefined) {
    profile = options.database;
  }
  const tempCredsFeatureEnabled = config.featureToggles.awsDatasourcesTempCredentials;
  const awsAssumeRoleEnabled = config.awsAssumeRoleEnabled;
  const awsAllowedAuthProviders = useMemo(
    () =>
      config.awsAllowedAuthProviders
        .filter((provider) => (provider === AwsAuthType.GrafanaAssumeRole ? tempCredsFeatureEnabled : true))
        .filter(isAwsAuthType),
    [tempCredsFeatureEnabled]
  );

  const currentProvider = awsAuthProviderOptions.find((p) => p.value === options.jsonData.authType);

  useEffect(() => {
    // Make sure a authType exists in the current model
    if (!currentProvider && awsAllowedAuthProviders.length) {
      onOptionsChange({
        ...options,
        jsonData: {
          ...options.jsonData,
          authType: awsAllowedAuthProviders[0],
        },
      });
    }
  }, [currentProvider, options, onOptionsChange, awsAllowedAuthProviders]);

  useEffect(() => {
    if (!loadRegions) {
      return;
    }

    loadRegions().then((regions) => setRegions(regions.map(toOption)));
  }, [loadRegions]);

  return (
    <FieldSet label={skipHeader ? '' : 'Connection Details'} data-testid="connection-config">
      <InlineField
        label="Authentication Provider"
        labelWidth={labelWidth}
        tooltip="Specify which AWS credentials chain to use."
      >
        <Select
          aria-label="Authentication Provider"
          className="width-30"
          value={currentProvider}
          options={awsAuthProviderOptions.filter((opt) => awsAllowedAuthProviders.includes(opt.value!))}
          defaultValue={options.jsonData.authType}
          onChange={(option) => {
            onUpdateDatasourceJsonDataOptionSelect(props, 'authType')(option);
          }}
          menuShouldPortal={true}
        />
      </InlineField>
      {options.jsonData.authType === 'credentials' && (
        <InlineField
          label="Credentials Profile Name"
          labelWidth={labelWidth}
          tooltip="Credentials profile name, as specified in ~/.aws/credentials, leave blank for default."
        >
          <Input
            aria-label="Credentials Profile Name"
            className="width-30"
            placeholder="default"
            value={profile}
            onChange={onUpdateDatasourceJsonDataOption(props, 'profile')}
          />
        </InlineField>
      )}

      {options.jsonData.authType === 'keys' && (
        <>
          <InlineField label="Access Key ID" labelWidth={labelWidth}>
            {props.options.secureJsonFields?.accessKey ? (
              <ButtonGroup className="width-30">
                <Input disabled placeholder="Configured" />
                <ToolbarButton
                  icon="edit"
                  tooltip="Edit Access Key ID"
                  type="button"
                  onClick={onUpdateDatasourceResetOption(props as any, 'accessKey')}
                />
              </ButtonGroup>
            ) : (
              <Input
                aria-label="Access Key ID"
                className="width-30"
                value={options.secureJsonData?.accessKey ?? ''}
                onChange={onUpdateDatasourceSecureJsonDataOption(props, 'accessKey')}
              />
            )}
          </InlineField>

          <InlineField label="Secret Access Key" labelWidth={labelWidth}>
            {props.options.secureJsonFields?.secretKey ? (
              <ButtonGroup className="width-30">
                <Input disabled placeholder="Configured" />
                <ToolbarButton
                  icon="edit"
                  type="button"
                  tooltip="Edit Secret Access Key"
                  onClick={onUpdateDatasourceResetOption(props as any, 'secretKey')}
                />
              </ButtonGroup>
            ) : (
              <Input
                aria-label="Secret Access Key"
                className="width-30"
                value={options.secureJsonData?.secretKey ?? ''}
                onChange={onUpdateDatasourceSecureJsonDataOption(props, 'secretKey')}
              />
            )}
          </InlineField>
        </>
      )}

      {awsAssumeRoleEnabled && (
        <>
          <InlineField
            label="Assume Role ARN"
            labelWidth={labelWidth}
            tooltip="Optionally, specify the ARN of a role to assume. Specifying a role here will ensure that the selected authentication provider is used to assume the specified role rather than using the credentials directly. Leave blank if you don't need to assume a role at all"
          >
            <Input
              aria-label="Assume Role ARN"
              className="width-30"
              placeholder="arn:aws:iam:*"
              value={options.jsonData.assumeRoleArn || ''}
              onChange={onUpdateDatasourceJsonDataOption(props, 'assumeRoleArn')}
            />
          </InlineField>
          {options.jsonData.authType !== AwsAuthType.GrafanaAssumeRole && (
            <InlineField
              label="External ID"
              labelWidth={labelWidth}
              tooltip="If you are assuming a role in another account, that has been created with an external ID, specify the external ID here."
            >
              <Input
                aria-label="External ID"
                className="width-30"
                placeholder="External ID"
                value={options.jsonData.externalId || ''}
                onChange={onUpdateDatasourceJsonDataOption(props, 'externalId')}
              />
            </InlineField>
          )}
        </>
      )}
      {!skipEndpoint && (
        <InlineField
          label="Endpoint"
          labelWidth={labelWidth}
          tooltip="Optionally, specify a custom endpoint for the service"
        >
          <Input
            aria-label="Endpoint"
            className="width-30"
            placeholder={props.defaultEndpoint ?? 'https://{service}.{region}.amazonaws.com'}
            value={options.jsonData.endpoint || ''}
            onChange={onUpdateDatasourceJsonDataOption(props, 'endpoint')}
          />
        </InlineField>
      )}
      <InlineField
        label="Default Region"
        labelWidth={labelWidth}
        tooltip="Specify the region, such as for US West (Oregon) use ` us-west-2 ` as the region."
      >
        <Select
          aria-label="Default Region"
          className="width-30"
          value={regions.find((region) => region.value === options.jsonData.defaultRegion)}
          options={regions}
          defaultValue={options.jsonData.defaultRegion}
          allowCustomValue={true}
          onChange={onUpdateDatasourceJsonDataOptionSelect(props, 'defaultRegion')}
          formatCreateLabel={(r) => `Use region: ${r}`}
          menuShouldPortal={true}
        />
      </InlineField>
      {props.children}
    </FieldSet>
  );
};
