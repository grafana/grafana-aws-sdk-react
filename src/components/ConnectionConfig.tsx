import React, { FC, useEffect, useMemo, useState } from 'react';
import { Input, Select, ButtonGroup, ToolbarButton, Collapse, useStyles2, Field } from '@grafana/ui';
import {
  DataSourcePluginOptionsEditorProps,
  onUpdateDatasourceJsonDataOptionSelect,
  onUpdateDatasourceResetOption,
  onUpdateDatasourceJsonDataOption,
  onUpdateDatasourceSecureJsonDataOption,
} from '@grafana/data';
import { config } from '@grafana/runtime';
import { standardRegions } from '../regions';
import { AwsAuthDataSourceJsonData, AwsAuthDataSourceSecureJsonData, AwsAuthType } from '../types';
import { awsAuthProviderOptions } from '../providers';
import { css } from '@emotion/css';
import { ConfigSection, ConfigSubSection } from '@grafana/experimental';

const DS_TYPES_THAT_SUPPORT_TEMP_CREDS = ['cloudwatch'];
const toOption = (value: string) => ({ value, label: value });
const isAwsAuthType = (value: any): value is AwsAuthType => {
  return typeof value === 'string' && awsAuthProviderOptions.some((opt) => opt.value === value);
};
export interface ConnectionConfigProps<
  J extends AwsAuthDataSourceJsonData = AwsAuthDataSourceJsonData,
  S = AwsAuthDataSourceSecureJsonData
> extends DataSourcePluginOptionsEditorProps<J, S> {
  standardRegions?: string[];
  loadRegions?: () => Promise<string[]>;
  defaultEndpoint?: string;
  skipHeader?: boolean;
  skipEndpoint?: boolean;
  children?: React.ReactNode;
  labelWidth?: number;
  inExperimentalAuthComponent?: boolean;
  externalId?: string;
}

export const ConnectionConfig: FC<ConnectionConfigProps> = (props: ConnectionConfigProps) => {
  const [isARNInstructionsOpen, setIsARNInstructionsOpen] = useState(false);
  const [regions, setRegions] = useState((props.standardRegions || standardRegions).map(toOption));
  const { loadRegions, onOptionsChange, skipHeader = false, skipEndpoint = false } = props;
  const { options } = props;
  let profile = options.jsonData.profile;
  if (profile === undefined) {
    profile = options.database;
  }
  const tempCredsFeatureEnabled =
    config.featureToggles.awsDatasourcesTempCredentials && DS_TYPES_THAT_SUPPORT_TEMP_CREDS.includes(options.type);
  const awsAssumeRoleEnabled = config.awsAssumeRoleEnabled ?? true;
  const awsAllowedAuthProviders = useMemo(
    () =>
      config.awsAllowedAuthProviders
        .filter((provider) => (provider === AwsAuthType.GrafanaAssumeRole ? tempCredsFeatureEnabled : true))
        .filter(isAwsAuthType),
    [tempCredsFeatureEnabled]
  );

  const currentProvider = awsAuthProviderOptions.find((p) => p.value === options.jsonData.authType);

  useEffect(() => {
    // Make sure a authType exists in the current model
    if (!currentProvider && awsAllowedAuthProviders.length) {
      onOptionsChange({
        ...options,
        jsonData: {
          ...options.jsonData,
          authType: awsAllowedAuthProviders[0],
        },
      });
    }
  }, [currentProvider, options, onOptionsChange, awsAllowedAuthProviders]);

  useEffect(() => {
    if (!loadRegions) {
      return;
    }

    loadRegions().then((regions) => setRegions(regions.map(toOption)));
  }, [loadRegions]);

  const styles = useStyles2(getStyles);

  return (
    <div data-testid="connection-config">
      <ConfigSection title={skipHeader ? '' : 'Connection Details'} data-testid="connection-config">
        <ConfigSubSection title="Authentication">
          <Field
            label="Authentication Provider"
            description="Specify which AWS credentials chain to use."
            htmlFor="authProvider"
          >
            <Select
              id="authProvider"
              aria-label="Authentication Provider"
              value={currentProvider}
              options={awsAuthProviderOptions.filter((opt) => awsAllowedAuthProviders.includes(opt.value!))}
              defaultValue={options.jsonData.authType}
              onChange={(option) => {
                onUpdateDatasourceJsonDataOptionSelect(props, 'authType')(option);
              }}
              menuShouldPortal={true}
            />
          </Field>
          {options.jsonData.authType === 'credentials' && (
            <Field
              label="Credentials Profile Name"
              description="Credentials profile name, as specified in ~/.aws/credentials, leave blank for default."
              htmlFor="credentialsProfileName"
            >
              <Input
                id="credentialsProfileName"
                aria-label="Credentials Profile Name"
                placeholder="default"
                value={profile}
                onChange={onUpdateDatasourceJsonDataOption(props, 'profile')}
              />
            </Field>
          )}
          {options.jsonData.authType === 'keys' && (
            <>
              <Field label="Access Key ID" htmlFor="accessKeyId">
                {props.options.secureJsonFields?.accessKey ? (
                  <ButtonGroup>
                    <Input disabled placeholder="Configured" id="accessKeyId" />
                    <ToolbarButton
                      icon="edit"
                      tooltip="Edit Access Key ID"
                      type="button"
                      onClick={onUpdateDatasourceResetOption(props as any, 'accessKey')}
                    />
                  </ButtonGroup>
                ) : (
                  <Input
                    id="accessKeyId"
                    aria-label="Access Key ID"
                    value={options.secureJsonData?.accessKey ?? ''}
                    onChange={onUpdateDatasourceSecureJsonDataOption(props, 'accessKey')}
                  />
                )}
              </Field>

              <Field label="Secret Access Key" htmlFor="secretKey">
                {props.options.secureJsonFields?.secretKey ? (
                  <ButtonGroup>
                    <Input disabled placeholder="Configured" />
                    <ToolbarButton
                      id="secretKey"
                      aria-label="Edit secret access key"
                      icon="edit"
                      type="button"
                      tooltip="Edit Secret Access Key"
                      onClick={onUpdateDatasourceResetOption(props as any, 'secretKey')}
                    />
                  </ButtonGroup>
                ) : (
                  <Input
                    id="secretKey"
                    aria-label="Secret Access Key"
                    value={options.secureJsonData?.secretKey ?? ''}
                    onChange={onUpdateDatasourceSecureJsonDataOption(props, 'secretKey')}
                  />
                )}
              </Field>
            </>
          )}
        </ConfigSubSection>

        <ConfigSubSection title="Assume Role">
          {options.jsonData.authType === AwsAuthType.GrafanaAssumeRole && (
            <div className={styles.assumeRoleInstructions}>
              <Collapse
                label={'How to create an IAM role for grafana to assume:'}
                collapsible={true}
                isOpen={isARNInstructionsOpen}
                onToggle={() => setIsARNInstructionsOpen(!isARNInstructionsOpen)}
              >
                <ol>
                  <li>
                    <p>
                      1. Create a new IAM role in the AWS console, and select <code>Another AWS account</code> as the{' '}
                      <code>Trusted entity</code>.
                    </p>
                  </li>
                  <li>
                    <p>
                      2. Enter the account ID of the Grafana account that has permission to assume this role:
                      <code> 008923505280 </code> and check the <code>Require external ID</code> box.
                    </p>
                  </li>
                  <li>
                    <p>
                      3. Enter the following external ID:{' '}
                      <code>{props.externalId || 'External Id is currently unavailable'}</code> and click{' '}
                      <code>Next</code>.
                    </p>
                  </li>
                  <li>
                    <p>
                      4. Add any required permissions you would like Grafana to be able to access on your behalf. For
                      more details on our permissions please{' '}
                      <a
                        href="https://grafana.com/docs/grafana/latest/datasources/aws-cloudwatch/"
                        target="_blank"
                        rel="noreferrer"
                      >
                        read through our documentation
                      </a>
                      .
                    </p>
                  </li>
                  <li>
                    <p>
                      5. Give the role a name and description, and click <code>Create role</code>.
                    </p>
                  </li>
                  <li>
                    <p>
                      6. Copy the ARN of the role you just created and paste it into the <code>Assume Role ARN</code>{' '}
                      field below.
                    </p>
                  </li>
                </ol>
              </Collapse>
            </div>
          )}
          {awsAssumeRoleEnabled && (
            <>
              <Field
                htmlFor="assumeRoleArn"
                label="Assume Role ARN"
                description="Optionally, specify the ARN of a role to assume. Specifying a role here will ensure that the
                  selected authentication provider is used to assume the role rather than using the
                  credentials directly."
              >
                <Input
                  id="assumeRoleArn"
                  aria-label="Assume Role ARN"
                  placeholder="arn:aws:iam:*"
                  value={options.jsonData.assumeRoleArn || ''}
                  onChange={onUpdateDatasourceJsonDataOption(props, 'assumeRoleArn')}
                />
              </Field>
              {options.jsonData.authType !== AwsAuthType.GrafanaAssumeRole && (
                <Field
                  htmlFor="externalId"
                  label="External ID"
                  description="If you are assuming a role in another account, that has been created with an external ID, specify the external ID here."
                >
                  <Input
                    id="externalId"
                    aria-label="External ID"
                    placeholder="External ID"
                    value={options.jsonData.externalId || ''}
                    onChange={onUpdateDatasourceJsonDataOption(props, 'externalId')}
                  />
                </Field>
              )}
            </>
          )}
        </ConfigSubSection>
        <ConfigSubSection title="Additional Settings">
          {!skipEndpoint && options.jsonData.authType !== AwsAuthType.GrafanaAssumeRole && (
            <Field
              label="Endpoint"
              description="Optionally, specify a custom endpoint for the service"
              htmlFor="endpoint"
            >
              <Input
                id="endpoint"
                aria-label="Endpoint"
                placeholder={props.defaultEndpoint ?? 'https://{service}.{region}.amazonaws.com'}
                value={options.jsonData.endpoint || ''}
                onChange={onUpdateDatasourceJsonDataOption(props, 'endpoint')}
              />
            </Field>
          )}

          <Field
            label="Default Region"
            description="Specify the region, such as for US West (Oregon) use ` us-west-2 ` as the region."
            htmlFor="defaultRegion"
          >
            <Select
              id="defaultRegion"
              aria-label="Default Region"
              value={regions.find((region) => region.value === options.jsonData.defaultRegion)}
              options={regions}
              defaultValue={options.jsonData.defaultRegion}
              allowCustomValue={true}
              onChange={onUpdateDatasourceJsonDataOptionSelect(props, 'defaultRegion')}
              formatCreateLabel={(r) => `Use region: ${r}`}
              menuShouldPortal={true}
            />
          </Field>
        </ConfigSubSection>
        {props.children}
      </ConfigSection>
    </div>
  );
};

function getStyles() {
  return {
    assumeRoleInstructions: css({
      maxWidth: '715px',
    }),
  };
}
