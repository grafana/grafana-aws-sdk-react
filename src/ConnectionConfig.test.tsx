import '@testing-library/jest-dom/extend-expect';
import React from 'react';
import { render, screen, waitFor } from '@testing-library/react';
import { AwsAuthDataSourceJsonData, AwsAuthDataSourceSecureJsonData, AwsAuthType } from './types';
import { ConnectionConfig, ConnectionConfigProps } from './ConnectionConfig';

const getProps = (propOverrides?: object) => {
  const props: ConnectionConfigProps<AwsAuthDataSourceJsonData, AwsAuthDataSourceSecureJsonData> = {
    options: {
      id: 21,
      typeName: 'aws',
      uid: '',
      orgId: 1,
      name: 'aws-plugin-name',
      type: 'aws',
      basicAuth: false,
      basicAuthUser: '',
      basicAuthPassword: '',
      withCredentials: false,
      isDefault: false,
      version: 1,
      readOnly: false,
      typeLogoUrl: '',
      url: '',
      password: '',
      access: '',
      database: '',
      user: '',
      jsonData: {
        authType: undefined,
        profile: '',
        assumeRoleArn: '',
        externalId: '',
        defaultRegion: '',
        endpoint: '',
      },
      secureJsonFields: {
        accessKey: false,
        secretKey: false,
        sessionToken: false,
      },
      secureJsonData: {
        accessKey: '',
        secretKey: '',
        sessionToken: '',
      },
    },
    onOptionsChange: jest.fn(),
  };

  Object.assign(props, propOverrides);

  return props;
};

const resetWindow = () => {
  (window as any).grafanaBootData = {
    settings: {
      awsAllowedAuthProviders: [AwsAuthType.EC2IAMRole, AwsAuthType.Keys],
      awsAssumeRoleEnabled: false,
    },
  };
};

describe('ConnectionConfig', () => {
  beforeEach(() => resetWindow());
  afterEach(() => resetWindow());

  it('should use auth type from props if its set', async () => {
    const onOptionsChange = jest.fn();
    const props = getProps({ onOptionsChange, options: { jsonData: { authType: AwsAuthType.Keys } } });
    render(<ConnectionConfig {...props} />);
    await waitFor(() => expect(screen.getByTestId('connection-config')).toBeInTheDocument());
    expect(screen.getByText('Access & secret key')).toBeInTheDocument();
    expect(onOptionsChange).not.toHaveBeenCalledWith();
  });

  it('should auto select first auth type if its not set', async () => {
    const onOptionsChange = jest.fn();
    const props = getProps({ onOptionsChange });
    render(<ConnectionConfig {...props} />);
    await waitFor(() => expect(screen.getByTestId('connection-config')).toBeInTheDocument());

    const config = props.options;
    expect(onOptionsChange).toHaveBeenCalledWith({
      ...config,
      jsonData: {
        ...config.jsonData,
        authType: AwsAuthType.EC2IAMRole,
      },
    });
  });

  it('should show secret field if auth type is keys', async () => {
    const props = getProps({ options: { jsonData: { authType: AwsAuthType.Keys } } });
    render(<ConnectionConfig {...props} />);
    await waitFor(() => expect(screen.getByTestId('connection-config')).toBeInTheDocument());

    expect(screen.getByText('Access Key ID')).toBeInTheDocument();
    expect(screen.getByText('Secret Access Key')).toBeInTheDocument();
  });

  it('should render endpoint if skipEndpoint prop is missing', async () => {
    const props = getProps();
    render(<ConnectionConfig {...props} />);
    await waitFor(() => expect(screen.getByTestId('connection-config')).toBeInTheDocument());
    expect(screen.getByText('Endpoint')).toBeInTheDocument();
  });

  it('should render endpoint if skipEndpoint prop is set to false', async () => {
    const props = getProps({ skipEndpoint: false });
    render(<ConnectionConfig {...props} />);
    await waitFor(() => expect(screen.getByTestId('connection-config')).toBeInTheDocument());
    expect(screen.queryByText('Endpoint')).toBeInTheDocument();
  });

  it('should not render endpoint if skipEndpoint prop is set to true', async () => {
    const props = getProps({ skipEndpoint: true });
    render(<ConnectionConfig {...props} />);
    await waitFor(() => expect(screen.getByTestId('connection-config')).toBeInTheDocument());
    expect(screen.queryByText('Endpoint')).not.toBeInTheDocument();
  });

  it('should render default header if skipHeader prop is missing', async () => {
    const props = getProps();
    render(<ConnectionConfig {...props} />);
    await waitFor(() => expect(screen.getByTestId('connection-config')).toBeInTheDocument());
    expect(screen.queryByText('Connection Details')).toBeInTheDocument();
  });

  it('should render default header if skipHeader prop is set to false', async () => {
    const props = getProps({ skipHeader: false });
    render(<ConnectionConfig {...props} />);
    await waitFor(() => expect(screen.getByTestId('connection-config')).toBeInTheDocument());
    expect(screen.queryByText('Connection Details')).toBeInTheDocument();
  });

  it('should not render default header if skipHeader prop is set to true', async () => {
    const props = getProps({ skipHeader: true });
    render(<ConnectionConfig {...props} />);
    await waitFor(() => expect(screen.getByTestId('connection-config')).toBeInTheDocument());
    expect(screen.queryByText('Connection Details')).not.toBeInTheDocument();
  });

  it('should use default auth if awsAllowedAuthProviders was not found on window obj', async () => {
    (window as any).grafanaBootData = {
      settings: {},
    };
    const onOptionsChange = jest.fn();
    const props = getProps({ onOptionsChange });
    render(<ConnectionConfig {...props} />);
    await waitFor(() => expect(screen.getByTestId('connection-config')).toBeInTheDocument());

    const config = props.options;
    expect(onOptionsChange).toHaveBeenCalledWith({
      ...config,
      jsonData: {
        ...config.jsonData,
        authType: AwsAuthType.Default,
      },
    });
  });

  it('should render assume role if awsAssumeRoleEnabled was not found on window obj', async () => {
    (window as any).grafanaBootData = {
      settings: {},
    };
    const props = getProps();
    render(<ConnectionConfig {...props} />);
    await waitFor(() => expect(screen.getByTestId('connection-config')).toBeInTheDocument());
    expect(screen.queryByText('Assume Role ARN')).toBeInTheDocument();
  });

  it('should not render assume role if awsAssumeRoleEnabled was set to false', async () => {
    (window as any).grafanaBootData = {
      settings: {
        awsAssumeRoleEnabled: false,
      },
    };
    const props = getProps();
    render(<ConnectionConfig {...props} />);
    await waitFor(() => expect(screen.getByTestId('connection-config')).toBeInTheDocument());
    expect(screen.queryByText('Assume Role ARN')).not.toBeInTheDocument();
  });

  it('should render assume role if awsAssumeRoleEnabled was set to true', async () => {
    (window as any).grafanaBootData = {
      settings: {
        awsAssumeRoleEnabled: true,
      },
    };
    const props = getProps();
    render(<ConnectionConfig {...props} />);
    await waitFor(() => expect(screen.getByTestId('connection-config')).toBeInTheDocument());
    expect(screen.queryByText('Assume Role ARN')).toBeInTheDocument();
  });
});
