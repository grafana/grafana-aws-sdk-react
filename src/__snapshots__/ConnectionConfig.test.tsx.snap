// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConnectionConfig should render component 1`] = `
<fieldset
  className="css-1qipljx"
  data-testid="connection-config"
>
  <legend
    className="css-xim8hk"
  >
    Connection Details
  </legend>
  <div
    className="css-1f5gueu"
  >
    <label
      className="css-knpcrf"
    >
      Authentication Provider
      <div
        className="css-1vzus6i-Icon"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        <svg
          data-testid="info-circle"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        />
      </div>
    </label>
    <div
      className="css-haofya-input-wrapper width-30 css-smioj7-Component"
      onKeyDown={[Function]}
    >
      <span
        aria-atomic="false"
        aria-live="polite"
        aria-relevant="additions text"
        className="css-1f43avz-a11yText-A11yText"
      />
      <div
        className="css-k6ld91"
        onMouseDown={[Function]}
        onTouchEnd={[Function]}
      >
        <div
          className="css-1kfautg-grafana-select-value-container"
        >
          <div
            className="css-sn2lc5"
          >
            Choose
          </div>
          <div
            className="css-17ragm3-Component"
          >
            <div
              className=""
              style={
                Object {
                  "display": "inline-block",
                }
              }
            >
              <input
                aria-autocomplete="list"
                autoCapitalize="none"
                autoComplete="off"
                autoCorrect="off"
                disabled={false}
                id="react-select-2-input"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                spellCheck="false"
                style={
                  Object {
                    "background": 0,
                    "border": 0,
                    "boxSizing": "content-box",
                    "color": "inherit",
                    "fontSize": "inherit",
                    "label": "input",
                    "opacity": 1,
                    "outline": 0,
                    "padding": 0,
                    "width": "1px",
                  }
                }
                tabIndex="0"
                type="text"
                value=""
              />
              <div
                style={
                  Object {
                    "height": 0,
                    "left": 0,
                    "overflow": "scroll",
                    "position": "absolute",
                    "top": 0,
                    "visibility": "hidden",
                    "whiteSpace": "pre",
                  }
                }
              >
                
              </div>
            </div>
          </div>
        </div>
        <div
          className="css-uvldi-input-suffix"
        >
          <div
            className="css-1vzus6i-Icon"
          >
            <svg
              data-testid="angle-down"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="css-1f5gueu"
  >
    <label
      className="css-knpcrf"
    >
      Endpoint
      <div
        className="css-1vzus6i-Icon"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        <svg
          data-testid="info-circle"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        />
      </div>
    </label>
    <div
      className="css-xcstkt-input-wrapper width-30"
    >
      <div
        className="css-1w5c5dq-input-inputWrapper"
      >
        <input
          className="css-1064hy6-input-input"
          onChange={[Function]}
          placeholder="https://{service}.{region}.amazonaws.com"
          style={
            Object {
              "paddingLeft": 12,
              "paddingRight": undefined,
            }
          }
          value=""
        />
      </div>
    </div>
  </div>
  <div
    className="css-1f5gueu"
  >
    <label
      className="css-knpcrf"
    >
      Default Region
      <div
        className="css-1vzus6i-Icon"
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        <svg
          data-testid="info-circle"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        />
      </div>
    </label>
    <div
      className="css-haofya-input-wrapper width-30 css-smioj7-Component"
      onKeyDown={[Function]}
    >
      <span
        aria-atomic="false"
        aria-live="polite"
        aria-relevant="additions text"
        className="css-1f43avz-a11yText-A11yText"
      />
      <div
        className="css-k6ld91"
        onMouseDown={[Function]}
        onTouchEnd={[Function]}
      >
        <div
          className="css-1kfautg-grafana-select-value-container"
        >
          <div
            className="css-sn2lc5"
          >
            Choose
          </div>
          <div
            className="css-17ragm3-Component"
          >
            <div
              className=""
              style={
                Object {
                  "display": "inline-block",
                }
              }
            >
              <input
                aria-autocomplete="list"
                autoCapitalize="none"
                autoComplete="off"
                autoCorrect="off"
                disabled={false}
                id="react-select-3-input"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                spellCheck="false"
                style={
                  Object {
                    "background": 0,
                    "border": 0,
                    "boxSizing": "content-box",
                    "color": "inherit",
                    "fontSize": "inherit",
                    "label": "input",
                    "opacity": 1,
                    "outline": 0,
                    "padding": 0,
                    "width": "1px",
                  }
                }
                tabIndex="0"
                type="text"
                value=""
              />
              <div
                style={
                  Object {
                    "height": 0,
                    "left": 0,
                    "overflow": "scroll",
                    "position": "absolute",
                    "top": 0,
                    "visibility": "hidden",
                    "whiteSpace": "pre",
                  }
                }
              >
                
              </div>
            </div>
          </div>
        </div>
        <div
          className="css-uvldi-input-suffix"
        >
          <div
            className="css-1vzus6i-Icon"
          >
            <svg
              data-testid="angle-down"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</fieldset>
`;
